src/cpp/session/modules/SessionCodeTools.R

```diff
@@ -1246,5 +1412,58 @@
 })
 
 .rs.addFunction("makePrimitiveWrapper", function(x) {
-   wrapper <- eval(parse(text = capture.output(x)), envir = parent.frame(1))
+   eval(parse(text = capture.output(x)), envir = parent.frame(1))
 })
+
+.rs.addFunction("extractNativeSymbols", function(DLL, collapse = TRUE)
+{
+   info <- getDLLRegisteredRoutines(DLL)
+   result <- lapply(info, function(routine) {
+      as.character(names(routine))
+   })
+   
+   if (collapse)
+      result <- as.character(unlist(result))
+   
+   result
+})
+
+.rs.addFunction("getNativeSymbols", function(package)
+{
+   loadedDLLs <- getLoadedDLLs()
+   if (package %in% names(loadedDLLs))
+      return(.rs.extractNativeSymbols(loadedDLLs[[package]]))
+   
+   reExtension <- paste("\\", .Platform$dynlib.ext, "$", sep = "")
+   
+   # Try loading the DLL temporarily so we can extract the symbols.
+   # Note that the shared object name does not necessarily match that
+   # of the package; e.g. `data.table` munges the object name.
+   libPath <- system.file("libs", package = package)
+   dllNames <- sub(
+      reExtension,
+      "",
+      list.files(libPath, pattern = reExtension)
+   )
+   
+   as.character(unlist(lapply(dllNames, function(name) {
+      
+      # TODO: Are there side-effects of this call that we want to avoid? If so
+      # we might want to execute this in a separate R process.
+      DLL <- try(
```

Comment:
There definitely can be side effects (rare but library dependency so not knowable). Also, you can't just do an unload in on.exit because that will pull it out from under anyone that was already using it (oh but wait, perhaps you only do this if the library wasn't already loaded, in which case it's fine). 

If it is true that you only load DLLs not already loaded then the side effects may be so rare/inaffectual that we don't need to worry about them.